draft for middleware interface

dirk-thomas opened the pull request:
@gerkey @esteve @tfoote @wjwwood fyi the current draft for the middleware interface whitepaper. 

I still have to write the last part describing the actually implemented architecture.

wjwwood commented on articles/60_ros_middleware_interface.md:
```diff
@@ -0,0 +1,236 @@
+    |      mw impl N       |   3) convert the DDS sample into a ROS message and invoke subscriber callback
+    +----------------------+
+
+Depending on the middleware implementation the extra conversion can be avoided by implementing serialization functions from as well as deserilization functions to the ROS messages.
```
This should probably either be:

```
Depending on the middleware implementation the extra conversion can be avoided by implementing serialization functions from, as well as deserilization functions to, the ROS messages.
```

Or

```
Depending on the middleware implementation the extra conversion can be avoided by implementing serialization functions from ROS messages as well as deserilization functions to the ROS messages.
```

wjwwood commented:
+1

dirk-thomas committed:
add doc for middleware interface

dirk-thomas commented:
I have updated the document and added the missing chapter. Please provide any kind of feedback so that we can announce it on the _ros-sig-ng-ros_ mailing list eaerly next week.

gerkey committed:
Spell check and some grammar edits.

gerkey commented:
+2.  Really nice explanation.

I made some cosmetic edits in b628668ad4c4d7b5b7425115194be0365746042a.

The announcement will separately include the description/video of the demo showing how easy the swap is, right?

wjwwood commented:
+1 for new stuff too.

tfoote commented:
+1 overall

I'd like to see a reminder that this is the middleware interface at the top of the design section L231 mentioning that this is the minimum layer enabling a developer to build a client library on top and is not expected to be a user facing API.

esteve commented:
+1 looks good to me too (including @gerkey's changes)

dirk-thomas commented:
@tfoote If you think the fact that it is not a user facing API is not clear enough then I would propose that it needs to be clarified way earlier then on L231.

In the first questions I tried to describe that the ROS client library is the user facing API. Also the following layer diagram show that. If you think it needs more clarification can you please propose a patch to the very beginning of the document?

tfoote committed:
clarifying reuse in client libraries.

dirk-thomas committed:
Merge pull request #24 from ros2/middleware_interface_clarification

clarifying reuse in client libraries.
